packages/@lwc/ssr-compiler/src/shared.ts

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (c) 2024, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+
+export type Expression = string;
+
+export type Instruction =
+    | IEmitTagName
+    | IEmitStaticString
+    | IEmitExpression
+    | IStartConditional
+    | IEndConditional
+    | IInvokeConnectedCallback
+    | IRenderChild
+    | IHoistImport
+    | IHoistInstantiation;
+
+export interface IEmitTagName {
+    kind: 'emitTagName';
+}
+
+export interface IEmitStaticString {
+    kind: 'emitStaticString';
+}
+
+export interface IEmitExpression {
+    kind: 'emitExpression';
+    expression: Expression;
+}
+
+export interface IStartConditional {
+    kind: 'startConditional';
+}
+
+export interface IEndConditional {
+    kind: 'endConditional';
+}
+
+export interface IInvokeConnectedCallback {
+    kind: 'invokeConnectedCallback';
+}
+
+export interface IRenderChild {
+    kind: 'renderChild';
+    dynamic: Expression | null;
+}
+
+export interface IHoistImport {
+    kind: 'hoistImport';
+}
+
+export interface IHoistInstantiation {
+    kind: 'hoistInstantiation';
+}
+
+export type GenerateMarkupFn = (
+    tagName: string,
+    props: Record<string, any> | null,
+    attrs: Record<string, any> | null,
+    slotted: Record<number | string, AsyncGenerator<string>> | null
```

Comment:
This was an intentional choice. Because generated code is interacting with other generated code and also with other runtimes, and because there isn't a mechanism to enforce the contracts through the type system, I opted to maintain the same number of arguments. It makes reading through the generated code much saner. There are only going to be one or two other codebases invoking these functions, so there isn't an ergonomics issue.